demos/dev_mode_key_provisioning/src/aws_dev_mode_key_provisioning.c

```diff
@@ -43,88 +44,486 @@
+        xEcPoint[ 0 ] = 0x04; /* Octet string. */
+        xEcPoint[ 1 ] = ( CK_BYTE ) xLength;
+
+        CK_ATTRIBUTE xPublicKeyTemplate[] =
+        {
+            { CKA_CLASS,     &xClass,           sizeof( xClass )                             },
+            { CKA_KEY_TYPE,  &xPublicKeyType,   sizeof( xPublicKeyType )                     },
+            { CKA_TOKEN,     &xTrue,            sizeof( xTrue )                              },
+            { CKA_VERIFY,    &xTrue,            sizeof( xTrue )                              },
+            { CKA_EC_PARAMS, xEcParams,         sizeof( xEcParams )                          },
+            { CKA_EC_POINT,  xEcPoint,          xLength + 2                                  },
+            { CKA_LABEL,     pucPublicKeyLabel, strlen( ( const char * ) pucPublicKeyLabel ) }
+        };
+
+        xResult = pxFunctionList->C_CreateObject( xSession,
+                                                  ( CK_ATTRIBUTE_PTR ) xPublicKeyTemplate,
+                                                  sizeof( xPublicKeyTemplate ) / sizeof( CK_ATTRIBUTE ),
+                                                  pxPublicKeyHandle );
+    }
+    else
+    {
+        xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+        configPRINTF( ( "Invalid key type. Supported options are CKK_RSA and CKK_EC" ) );
+    }
+
+    mbedtls_pk_free( &xMbedPkContext );
+
+    return xResult;
+}
+
+
 /*-----------------------------------------------------------*/
 
-/** @brief Converts PEM documents into DER formatted byte arrays.
- * This is a helper function from mbedTLS util pem2der.c
- * (https://github.com/ARMmbed/mbedtls/blob/development/programs/util/pem2der.c#L75)
- *
- * \param[in] pucInput          Pointer to PEM object
- * \param[in] xLen              Length of PEM object
- * \param[out] pucOutput        Pointer to buffer where DER object will be placed
- * \param[in/out] pxOlen        Pointer to length of DER buffer.  This value is updated
- *                              to contain the actual length of the converted DER object.
- *
- * \return 0 if successful.  Negative if conversion failed.  If buffer is not
- * large enough to hold converted object, pxOlen is still updated but -1 is returned.
- *
- */
+/* NOTE: C_GenerateKeyPair for RSA keys is not supported by the Amazon FreeRTOS mbedTLS PKCS #11 port. */
+CK_RV xProvisionGenerateKeyPairRSA( CK_SESSION_HANDLE xSession,
+                                    uint8_t * pucPrivateKeyLabel,
+                                    uint8_t * pucPublicKeyLabel,
+                                    CK_OBJECT_HANDLE_PTR pxPrivateKeyHandle,
+                                    CK_OBJECT_HANDLE_PTR pxPublicKeyHandle )
+{
+    CK_RV xResult;
+    CK_MECHANISM xMechanism =
+    {
+        CKM_RSA_PKCS_KEY_PAIR_GEN, NULL_PTR, 0
+    };
+    CK_FUNCTION_LIST_PTR pxFunctionList;
+    CK_ULONG xModulusBits = 2048;
```

Comment:
Could this instead be based on the #defines you have above for key byte field lengths (i.e. multiplied by eight)? That would make it easier to change the whole implementation over to a different modulus size.

Reply:
added a define for 2048.  still have some of the length parameters independently defined... I feel like I need to do more research into the causal relationship between these parameters.